crates/wg_network/src/routing.rs

```diff
@@ -7,3 +11,182 @@ pub struct SourceRoutingHeader {
+        !self.is_empty_route() && self.hop_index == self.hops.len() - 1
+    }
+    /// Returns true if the hop index is valid.
+    pub fn valid_hop_index(&self) -> bool {
+        !self.is_empty_route() && self.hop_index < self.hops.len()
+    }
+
+    // HOPS MANIPULATION
+    /// Appends a hop to the route.
+    pub fn append_hop(&mut self, hop: NodeId) {
+        self.hops.push(hop);
+    }
+
+    // WHOLE ROUTE MANIPULATION
+    /// Reverses the route.
+    pub fn reverse(&mut self) {
+        if self.is_empty_route() {
+            return;
+        }
+        self.hops.reverse();
+        self.hop_index = self.hops.len() - self.hop_index - 1;
+    }
+    /// Returns the reversed route.
+    pub fn get_reversed(&self) -> SourceRoutingHeader {
+        let mut clone = self.clone();
+        clone.reverse();
+        clone
+    }
+    /// Extracts a sub-route from the route.
+    /// If the range is decreasing, it also reverses the sub-route.
+    pub fn sub_route(&self, range: impl RangeBounds<usize>) -> Option<SourceRoutingHeader> {
+        let raw_start = match range.start_bound() {
+            Bound::Included(&start) => start,
+            Bound::Excluded(&start) => start,
+            Bound::Unbounded => 0,
+        };
+        let raw_end = match range.end_bound() {
+            Bound::Included(&end) => end,
+            Bound::Excluded(&end) => end,
+            Bound::Unbounded => self.hops.len(),
+        };
+
+        if raw_start > raw_end {
+            let start = match range.end_bound() {
+                Bound::Included(&end) => end,
+                Bound::Excluded(&end) => end + 1,
+                Bound::Unbounded => self.hops.len(),
+            };
+
+            let end = match range.start_bound() {
+                Bound::Included(&start) => start + 1,
+                Bound::Excluded(&start) => start,
+                Bound::Unbounded => 0,
+            };
+
+            if end > self.hops.len() {
+                return None;
+            }
+
+            let mut res = SourceRoutingHeader {
+                hop_index: self.hop_index.max(start) - start,
+                hops: self.hops[start..end].to_vec(),
+            };
+
+            res.reverse();
+
+            Some(res)
+        } else {
+            let start = match range.start_bound() {
+                Bound::Included(&start) => start,
+                Bound::Excluded(&start) => start + 1,
+                Bound::Unbounded => 0,
+            };
+
+            let end = match range.end_bound() {
+                Bound::Included(&end) => end + 1,
+                Bound::Excluded(&end) => end,
+                Bound::Unbounded => self.hops.len(),
+            };
+
+            if end > self.hops.len() {
+                return None;
+            }
+
+            Some(SourceRoutingHeader {
+                hop_index: self.hop_index.max(start) - start,
+                hops: self.hops[start..end].to_vec(),
+            })
+        }
+    }
+}
```

Comment:
Isn't there any more compact way to do this?

Reply:
I know it's weird, if you can figure it out please help me

Reply:
Maybe use Range https://doc.rust-lang.org/std/ops/struct.Range.html?

Reply:
```rust
fn main(){
    let numbers = vec![1, 2, 3, 4, 5, 6, 7, 8, 9];
    let subvec = subvector(&numbers, 2..5);
    println!("{:?}", subvec);
}

fn subvector<T: Clone> (v: &Vec<T>, range: Range<usize>) -> Vec<T>{
    v[range].to_vec()
}
```

Reply:
```suggestion
pub fn sub_route(&self, range: impl Range) -> Option<SourceRoutingHeader> {
    Some(SourceRoutingHeader{
        hops: self.self.hops.get(range)?.to_vec(),
        hop_index: self.hop_index.max(range.end - 1) - range.start,
    }
}
```
I put it down fast so i'm not sure it work.

Reply:
I knew I could do it that way, but the additional complexity is justified by the possibility of doing route.subroute(2..=0) that also reverses the route.
If you think it isn't useful, I'll simplify it.

Reply:
More than not useful. I honestly hadn't understood that. I was thinking it just checked boundaries.
I think it is not necessary when we have rev. You can always make another function that does both.